src/main/scala/is/hail/expr/ir/functions/Functions.scala

```diff
@@ -0,0 +1,148 @@
+package is.hail.expr.ir.functions
+
+import is.hail.asm4s._
+import is.hail.expr.{AST, FunType, Lambda, TypeTag}
+import is.hail.expr.ir._
+import is.hail.expr.types._
+import is.hail.utils._
+import is.hail.asm4s.coerce
+
+import scala.collection.mutable
+import scala.reflect.ClassTag
+
+object IRFunctionRegistry {
+
+  val registry: mutable.Map[String, Seq[(TypeTag, Seq[IR] => IR)]] = mutable.Map().withDefaultValue(Seq.empty)
+
+  def addIRFunction(f: IRFunction) {
+    val l = registry(f.name)
+    registry.put(f.name,
+      l :+ (f.argsTypeTag, { args: Seq[IR] =>
+        ApplyFunction(f, args)
+      }))
+  }
+
+  def addIR(name: String, types: Seq[Type], f: Seq[IR] => IR) {
+    val l = registry(name)
+    registry.put(name, l :+ ((FunType(types: _*), f)))
+  }
+
+  def lookupFunction(name: String, args: TypeTag): Option[Seq[IR] => IR] = {
```

Comment:
Oh, TypeTag, I got slightly confused.  I think you should just have a Seq[Type], we shouldn't distinguish between method and function in the IR (method is just a function with an extra argument) and TypeTag should go the was of AST.